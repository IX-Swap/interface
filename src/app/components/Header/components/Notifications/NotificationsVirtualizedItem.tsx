import React from 'react'
import { ListChildComponentProps } from 'react-window'
import { NotificationView } from 'app/components/Header/components/Notifications/NotificationView/NotificationView'
import { useVirtualizedListItemHelpers } from 'app/components/Header/components/Notifications/VirtualizedList/hooks/useVirtualizedListHelpers'

export interface NotificationsVirtualizedItemProps
  extends ListChildComponentProps {}

export const NotificationsVirtualizedItem = (
  props: NotificationsVirtualizedItemProps
) => {
  const { data, index, ...rest } = props
  const { ref } = useVirtualizedListItemHelpers(index, 0)
  const item = data[index]

  return (
    <div {...rest}>
      <div ref={ref}>
        <NotificationView data={item} />
      </div>
    </div>
  )
}
