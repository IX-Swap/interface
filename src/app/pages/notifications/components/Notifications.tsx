import React from 'react'
import { Grid } from '@material-ui/core'
import { PageHeader } from 'app/components/PageHeader/PageHeader'
import { MarkAllAsRead } from 'app/pages/notifications/components/MarkAllAsRead'
import { VSpacer } from 'components/VSpacer'
import { NotificationsList } from 'app/pages/notifications/components/NotificationsList'

export const Notifications = () => {
  return (
    <Grid container direction='column'>
      <Grid item>
        <PageHeader label='Notifications' alignment='flex-start' />
      </Grid>
      <Grid item container justify='flex-end'>
        <MarkAllAsRead />
      </Grid>
      <VSpacer size='small' />
      <Grid item>
        <NotificationsList />
      </Grid>
    </Grid>
  )
}
