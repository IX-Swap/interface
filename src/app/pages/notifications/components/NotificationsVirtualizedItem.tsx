import React from 'react'
import { ListChildComponentProps } from 'react-window'
import { useVirtualizedListItemHelpers } from 'components/VirtualizedList/hooks/useVirtualizedListHelpers'
import { MarkAsRead } from './MarkAsRead'
import { NotificationView } from './NotificationView'

export interface NotificationsVirtualizedItemProps
  extends ListChildComponentProps {}

export const NotificationsVirtualizedItem = (
  props: NotificationsVirtualizedItemProps
) => {
  const { data, index, ...rest } = props
  const { ref } = useVirtualizedListItemHelpers(index, 0)
  const item = data[index]
  return (
    <div {...rest}>
      <div ref={ref} style={{ marginBottom: -1 }}>
        <NotificationView data={item} action={<MarkAsRead data={item} />} />
      </div>
    </div>
  )
}
