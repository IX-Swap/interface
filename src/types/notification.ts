import { AppFeature } from './app'

export type NotificationType =
  | 'success'
  | 'error'
  | 'info'
  | 'warning'
  | 'BUY'
  | 'SELL'

export interface Notification {
  _id: string
  createdBy: string
  createdAt: string
  to: string
  service: string
  feature: AppFeature
  resourceId: string
  subject: string
  type: NotificationType
  message: string
  read: boolean
  readAt: string
  comment?: string
}
