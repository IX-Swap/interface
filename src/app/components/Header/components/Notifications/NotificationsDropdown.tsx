import React from 'react'
import { Grid } from '@mui/material'
import { Dropdown } from 'app/components/Dropdown/Dropdown'
import { AppRouterLink } from 'components/AppRouterLink'
import { NotificationsDropdownTrigger } from 'app/components/Header/components/Notifications/NotificationsDropdownTrigger/NotificationsDropdownTrigger'
import { NotificationsDropdownContent } from 'app/components/Header/components/Notifications/NotificationsDropdownContent/NotificationsDropdownContent'
import { useAppBreakpoints } from 'hooks/useAppBreakpoints'
import { AppRoute } from 'app/router/config'

export const NotificationsDropdown = () => {
  const { isTablet } = useAppBreakpoints()

  if (isTablet) {
    return (
      <AppRouterLink to={AppRoute.notifications}>
        <NotificationsDropdownTrigger />
      </AppRouterLink>
    )
  }

  return (
    <Grid item>
      <Dropdown
        trigger={NotificationsDropdownTrigger}
        // TODO Needs to add new content after complete design for notifications dropdown
        content={NotificationsDropdownContent}
      />
    </Grid>
  )
}
