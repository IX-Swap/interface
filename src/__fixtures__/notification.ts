import { AppFeature } from 'v2/types/app'
import { Notification } from 'v2/types/notification'

export const notification: Notification = {
  _id: '1',
  createdBy: '',
  createdAt: '2020-10-14T03:54:06.977Z',
  to: '',
  service: '',
  feature: AppFeature.Authentication,
  resourceId: '',
  subject: 'Cash Withdrawl',
  type: 'success',
  message: 'Your cash withdrawl request has been declined.',
  read: false,
  readAt: ''
}
