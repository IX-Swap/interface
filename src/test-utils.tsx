import React from 'react'
import { Router } from 'react-router-dom'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import {
  createGenerateClassName,
  StylesProvider
} from '@material-ui/core/styles'
import { ThemeProvider } from '@material-ui/styles'
import Themes from './v2/themes'
import { history } from 'v2/history'
import { UserProvider } from 'v2/auth/context'
import { UserStore } from 'v2/auth/context/store'
import { PasswordResetProvider } from 'v2/auth/context/password-reset'
import PasswordResetStore from 'v2/auth/context/password-reset/store'
import { PasswordResetStep } from 'v2/auth/context/password-reset/types'
import { AuthorizerTableStore } from 'v2/app/pages/authorizer/context/store'
import { AuthorizerTableStoreProvider } from 'v2/app/pages/authorizer/context'
import { DepositStore } from 'v2/app/pages/accounts/pages/banks/context/store'
import { DepositStoreProvider } from 'v2/app/pages/accounts/pages/banks/context'
import { ServicesProvider } from 'v2/services/useServices'
import { renderHook, RenderHookResult } from '@testing-library/react-hooks'
import { BreadcrumbsProvider } from 'v2/hooks/useBreadcrumbs'

const generateClassName = createGenerateClassName({
  productionPrefix: 'ix'
})

const BaseProviders: React.FC = ({ children }) => {
  return (
    <StylesProvider generateClassName={generateClassName}>
      <ThemeProvider theme={Themes.default}>
        <BreadcrumbsProvider>
          <Router history={history}>{children}</Router>
        </BreadcrumbsProvider>
      </ThemeProvider>
    </StylesProvider>
  )
}

const customRenderer = (
  ui: any,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => {
  return render(ui, { wrapper: BaseProviders, ...options })
}

export const testComponentId = 'test-component'

export const fakeUserStore: Partial<UserStore> = {
  activeTab: 0,
  setActiveTab: jest.fn()
}

export const renderWithUserStore = (
  ui: any,
  store?: Partial<UserStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <UserProvider value={{ ...fakeUserStore, ...store }}>
        {children}
      </UserProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const fakePasswordResetStore: Partial<PasswordResetStore> = {
  currentStep: PasswordResetStep.Request,
  setCurrentStep: jest.fn()
}

export const renderWithPasswordResetStore = (
  ui: any,
  store?: Partial<PasswordResetStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <PasswordResetProvider value={{ ...fakePasswordResetStore, ...store }}>
        {children}
      </PasswordResetProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const fakeAuthorizerTableStore: Partial<AuthorizerTableStore> = {
  idKey: '__',
  uri: '/api/source',
  _getItemId: jest.fn()
}

export const renderWithAuthorizerTableStore = (
  ui: any,
  store?: Partial<AuthorizerTableStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <AuthorizerTableStoreProvider
        value={{ ...fakeAuthorizerTableStore, ...store }}
      >
        {children}
      </AuthorizerTableStoreProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const fakeDepositStore: Partial<DepositStore> = {}

export const renderWithDepositStore = (
  ui: any,
  store?: Partial<DepositStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <DepositStoreProvider value={{ ...fakeDepositStore, ...store }}>
        {children}
      </DepositStoreProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const renderHookWithServiceProvider = (
  hookFn: any,
  store: object = {}
): RenderHookResult<any, any> => {
  const WithServiceProvider: React.FC = ({ children }) => (
    <ServicesProvider value={store}>{children}</ServicesProvider>
  )

  return renderHook(hookFn, { wrapper: WithServiceProvider })
}

export * from '@testing-library/react'
export { customRenderer as render }
