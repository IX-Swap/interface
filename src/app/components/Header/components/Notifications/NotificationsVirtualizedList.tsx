import React from 'react'
import { VirtualizedList } from 'app/components/Header/components/Notifications/VirtualizedList/VirtualizedList'
import { useNotifications } from 'app/pages/notifications/hooks/useNotifications'
import { NotificationsVirtualizedItem } from 'app/components/Header/components/Notifications/NotificationsVirtualizedItem'

export const NotificationsVirtualizedList = () => {
  const { data } = useNotifications()

  return (
    <VirtualizedList
      itemData={data}
      itemCount={data?.length}
      width='100%'
      itemComponent={NotificationsVirtualizedItem}
    />
  )
}
