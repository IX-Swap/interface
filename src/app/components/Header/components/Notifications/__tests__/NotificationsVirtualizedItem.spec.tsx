import { NotificationsVirtualizedItem } from 'app/components/Header/components/Notifications/NotificationsVirtualizedItem'
import React from 'react'
import { ListChildComponentProps } from 'react-window'
import { render } from 'test-utils'
import { notification } from '__fixtures__/notification'

jest.mock('helpers/dates', () => ({
  getTimeAgo: jest.fn(() => 'Now')
}))

describe('NotificationsVirtualizedItem', () => {
  const props: ListChildComponentProps = {
    data: [notification],
    index: 0,
    style: {}
  }

  afterAll(async () => {
    jest.clearAllMocks()
  })

  it('should match snapshot', () => {
    const { container } = render(<NotificationsVirtualizedItem {...props} />)
    expect(container).toMatchSnapshot()
  })
})
