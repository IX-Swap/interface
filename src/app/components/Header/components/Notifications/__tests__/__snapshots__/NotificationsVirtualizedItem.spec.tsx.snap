// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsVirtualizedItem should match snapshot 1`] = `
<div>
  <div>
    <div>
      <div
        class="MuiGrid-root makeStyles-container-1 css-vj1n65-MuiGrid-root"
        data-testid="wrapper"
      >
        <div
          class="MuiGrid-root MuiGrid-container makeStyles-inner-2 css-11lq3yg-MuiGrid-root"
          data-testid="notification-inner"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          />
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-zeroMinWidth makeStyles-content-3 css-175wz9v-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap makeStyles-title-4 css-1kaswgd-MuiTypography-root"
            >
              Cash Withdrawl
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-3phns2-MuiTypography-root"
            >
              Your cash withdrawl request has been declined.
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap fs-exclude css-1kaswgd-MuiTypography-root"
              style="color: rgb(52, 168, 123);"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column makeStyles-actions-5 css-vnie10-MuiGrid-root"
          >
            <div
              aria-disabled="true"
              aria-label="Mark as read"
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root Mui-disabled makeStyles-container-6 makeStyles-container-7 css-slzo6x-MuiButtonBase-root-MuiButton-root"
              data-mui-internal-clone-element="true"
              role="button"
              tabindex="-1"
            />
            <span
              class="MuiTypography-root MuiTypography-caption css-1fbb6wu-MuiTypography-root"
            >
              553 d
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
